File: ruby/corvy_sdk.rb
# CorvyBot SDK - v1.5.1
# Client library for building Corvy bots

require 'net/http'
require 'uri'
require 'json'
require 'time'
require 'shellwords'

# Type annotation helper for greedy parameters
class Greedy
  # Marker class for greedy parameters
end

# Helper for casting command arguments to the correct type
def cast_type(type, raw)
  case type
  when :string
    raw
  when :integer
    raw.to_i
  when :float
    raw.to_f
  when :boolean
    %w[1 true yes y t].include?(raw.downcase)
  else
    raise "Unsupported type: #{type}"
  end
end

# Parse arguments for a command
def parse_args(command_info, input_str, message)
  # Simple case - just pass message to handler
  return [message] if command_info[:params].empty?
  
  tokens = Shellwords.split(input_str)
  out_args = []
  idx = 0
  message_injected = false
  
  command_info[:params].each_with_index do |param, i|
    # If parameter is a message object
    if param[:type] == :message
      if message_injected
        # Second message not allowed unless it's optional
        if param[:optional]
          out_args << nil
          next
        end
        raise "Multiple Message parameters not allowed: #{param[:name]}"
      end
      out_args << message
      message_injected = true
      next
    end
    
    # Handle greedy parameter
    if param[:greedy]
      needed_for_rest = command_info[:params].length - (i + 1)
      take = [0, tokens.length - idx - needed_for_rest].max
      raw = tokens[idx, take].join(' ')
      idx += take
      out_args << cast_type(param[:type], raw)
      next
    end
    
    # Handle regular parameter
    if idx >= tokens.length
      if param[:default]
        out_args << param[:default]
        next
      end
      if param[:optional]
        out_args << nil
        next
      end
      raise "Missing value for parameter '#{param[:name]}'"
    end
    
    raw = tokens[idx]
    idx += 1
    
    if param[:optional] && raw.downcase == 'none'
      out_args << nil
    else
      out_args << cast_type(param[:type], raw)
    end
  end
  
  out_args
end

# User representation
class MessageUser
  attr_reader :id, :username, :is_bot
  
  def initialize(id, username, is_bot)
    @id = id
    @username = username
    @is_bot = is_bot
  end
end

# Message representation
class Message
  attr_reader :id, :content, :flock_name, :flock_id, :nest_name, :nest_id, :created_at, :user
  
  def initialize(id, content, flock_name, flock_id, nest_name, nest_id, created_at, user)
    @id = id
    @content = content
    @flock_name = flock_name
    @flock_id = flock_id
    @nest_name = nest_name
    @nest_id = nest_id
    @created_at = created_at
    @user = user
  end
end

class CorvyBot
  # Create a new bot instance
  # @param token [String] Token for the Corvy API
  # @param global_prefix [String] The prefix for all commands
  # @param api_base_url [String] The URL for the Corvy API
  # @param api_path [String] API path endpoint
  def initialize(token, global_prefix = "!", api_base_url = "https://corvy.chat", api_path = "/api/v1")
    @commands = {}
    @token = token
    @global_prefix = global_prefix
    @api_base_url = api_base_url
    @api_path = api_path
    @current_cursor = 0
    @events = {}
    
    # Set up signal handlers for graceful shutdown
    Signal.trap("INT") { shutdown }
  end
  
  # Register a command
  # @param prefix [String, nil] The prefix of the command
  # @return [Proc] Decorator function to register the command
  def command(prefix = nil)
    lambda do |func|
      command_name = prefix || "#{@global_prefix}#{func.name}"
      params = []
      
      # Extract parameter info from method definition if possible
      method_params = func.parameters
      method_params.each do |param_type, param_name|
        # Simple mapping for demonstration
        param_info = {
          name: param_name,
          type: :string,
          optional: param_type == :opt || param_type == :key,
          greedy: false,
          default: nil
        }
        
        # Special case for message parameter
        if param_name.to_s == 'message'
          param_info[:type] = :message
        end
        
        params << param_info
      end
      
      @commands[command_name] = {
        handler: func,
        params: params
      }
      
      func
    end
  end
  
  # Register an event handler
  # @param event [String, nil] The event to register for
  # @return [Proc] Decorator function to register the event
  def event(event = nil)
    lambda do |func|
      event_name = event || func.name
      @events[event_name] ||= []
      @events[event_name] << func
      func
    end
  end
  
  # Start the bot and begin processing messages
  def start
    begin
      puts "Starting bot..."
      
      # Authenticate first
      auth_response = make_request(:post, "#{@api_path}/auth")
      puts "Bot authenticated: #{auth_response["bot"]["name"]}"
      
      # Establish baseline (gets highest message ID but no messages)
      puts "Establishing baseline with server..."
      baseline_response = make_request(:get, "#{@api_path}/messages", cursor: 0)
      
      # Save the cursor for future requests
      if baseline_response["cursor"]
        @current_cursor = baseline_response["cursor"]
        puts "Baseline established. Starting with message ID: #{@current_cursor}"
      end
      
      # Log command prefixes
      command_prefixes = @commands.keys
      puts "Listening for commands: #{command_prefixes.join(", ")}"
      
      # Start processing messages
      process_message_loop
      
    rescue => e
      puts "Failed to start bot: #{e.message}"
      exit(1)
    end
  end
  
  # Send a message
  # @param flock_id [String, Integer] Flock ID
  # @param nest_id [String, Integer] Nest ID
  # @param content [String] Message content
  def send_message(flock_id, nest_id, content)
    begin
      puts "Sending message: \"#{content}\""
      
      make_request(:post, "#{@api_path}/flocks/#{flock_id}/nests/#{nest_id}/messages", nil, content: content)
      
    rescue => e
      puts "Failed to send message: #{e.message}"
    end
  end
  
  private
  
  # Process messages in a loop
  def process_message_loop
    loop do
      begin
        # Get new messages
        response = make_request(:get, "#{@api_path}/messages", cursor: @current_cursor)
        
        # Update cursor
        @current_cursor = response["cursor"] if response["cursor"]
        
        # Process each new message
        response["messages"]&.each do |msg|
          # Convert to Message object
          user = MessageUser.new(
            msg["user"]["id"],
            msg["user"]["username"],
            msg["user"]["is_bot"]
          )
          
          message = Message.new(
            msg["id"],
            msg["content"],
            msg["flock_name"],
            msg["flock_id"],
            msg["nest_name"],
            msg["nest_id"],
            Time.parse(msg["created_at"]),
            user
          )
          
          # Run on_message_raw events
          if @events["on_message_raw"]
            @events["on_message_raw"].each do |event_handler|
              event_handler.call(message)
            end
          end
          
          # Skip bot messages
          next if message.user.is_bot
          
          puts "Message from #{message.user.username} in #{message.flock_name}/#{message.nest_name}: #{message.content}"
          
          # Check for commands
          was_command = handle_command(message)
          
          # If it was a command, skip
          next if was_command
          
          # Run on_message events
          if @events["on_message"]
            @events["on_message"].each do |event_handler|
              event_handler.call(message)
            end
          end
        end
        
        # Wait before checking again
        sleep(1)
        
      rescue => e
        puts "Error fetching messages: #{e.message}"
        sleep(5) # Longer delay on error
      end
    end
  end
  
  # Handle command messages
  # @param message [Message] Message object
  # @return [Boolean] Whether a command was handled
  def handle_command(message)
    message_content = message.content.downcase
    
    # Check each command prefix
    @commands.each do |prefix, command_info|
      if message_content.start_with?(prefix.downcase)
        puts "Command detected: #{prefix}"
        
        begin
          # Parse arguments and call handler
          args = parse_args(
            command_info,
            message.content.sub(prefix, "").strip,
            message
          )
          
          response_content = command_info[:handler].call(*args)
          
          # Send the response
          send_message(message.flock_id, message.nest_id, response_content)
          
          # Return true to indicate command was handled
          return true
        rescue => e
          # Run on_command_exception events if available
          if @events["on_command_exception"]
            @events["on_command_exception"].each do |event_handler|
              event_handler.call(prefix, message, e)
            end
          end
          return false
        end
      end
    end
    
    # No commands were handled
    false
  end
  
  # Make an HTTP request to the Corvy API
  # @param method [Symbol] HTTP method (:get, :post, etc.)
  # @param path [String] API endpoint path
  # @param params [Hash] Query parameters
  # @param body [Hash] Request body
  # @return [Hash] Parsed JSON response
  def make_request(method, path, params = nil, body = nil)
    uri = URI.parse("#{@api_base_url}#{path}")
    uri.query = URI.encode_www_form(params) if params
    
    http = Net::HTTP.new(uri.host, uri.port)
    http.use_ssl = uri.scheme == 'https'
    
    request = case method
      when :get then Net::HTTP::Get.new(uri)
      when :post then Net::HTTP::Post.new(uri)
      when :put then Net::HTTP::Put.new(uri)
      when :delete then Net::HTTP::Delete.new(uri)
      else raise "Unsupported HTTP method: #{method}"
    end
    
    request["Authorization"] = "Bearer #{@token}"
    request["Content-Type"] = 'application/json'
    request.body = body.to_json if body
    
    response = http.request(request)
    
    if response.code.to_i >= 200 && response.code.to_i < 300
      JSON.parse(response.body)
    else
      raise "HTTP Error #{response.code}: #{response.body}"
    end
  end
  
  # Handle graceful shutdown
  def shutdown
    puts "Bot shutting down..."
    exit(0)
  end
end 

File: python/corvy_sdk.py
# CorvyBot SDK - v1.3.0
# Client library for building Corvy bots

import types
import aiohttp
import traceback
import json
import signal
import sys
import asyncio
from dataclasses import dataclass
from datetime import datetime
from typing import Awaitable, List, Dict, Callable, Any, Optional, Union, get_origin, get_args, Generic, TypeVar
import shlex
import inspect

__all__ = ["Greedy", "MessageUser", "Message", "CorvyBot"]

T = TypeVar("T")

class Greedy(Generic[T]):
    """Marker for greedy consumption of tokens."""
    pass

def cast_type(typ: type, raw: str) -> Any:
    if typ is str:
        return raw
    if typ is int:
        return int(raw)
    if typ is float:
        return float(raw)
    if typ is bool:
        return raw.lower() in ("1", "true", "yes", "y", "t")
    raise ValueError(f"Unsupported type: {typ!r}")

def is_union_type(ann):
    """"""
    return (
        get_origin(ann) is Union
        or isinstance(ann, types.UnionType)  # for Python 3.10's X|Y
    )

@dataclass
class MessageUser:
    id: int
    username: str
    is_bot: bool
    
@dataclass
class Message:
    id: int
    content: str
    flock_name: str
    flock_id: int
    nest_name: str
    nest_id: int
    created_at: datetime
    user: MessageUser

class CorvyBot:
    """
    Client library for building Corvy bots
    """
    
    def __init__(self, token: str, global_prefix: str = "!", api_base_url: str = "https://corvy.chat", api_path: str = "/api/v1"):
        """
        Create a new bot instance
        
        Args:
            token: Token for the Corvy API.
            global_prefix: The prefix for all commands. Defaults to an exclamation mark.
            api_base_url: The URL for the Corvy API.
        """
        self.commands: dict[str, Callable] = {}
        self.token = token
        self.global_prefix = global_prefix
        self.api_base_url = api_base_url
        self.api_path = api_path
        self.current_cursor = 0
        self.headers = {
            'Authorization': f"Bearer {token}",
            'Content-Type': 'application/json'
        }
        self.client_session: aiohttp.ClientSession | None = None
        self.events: dict[str, list[Awaitable]] = {}
        
        # Setup signal handler for graceful shutdown
        signal.signal(signal.SIGINT, self._handle_shutdown_stub)
    
    def command(self, prefix: str | None = None):
        """Register a command.
        
        Args:
            prefix: The prefix of the command. Defaults to the name of the function with the global prefix beforehand."""
            
        def _decorator_inst(func: Awaitable):
            self.commands[prefix or f"{self.global_prefix}{getattr(func, '__name__', None)}"] = func
            return func # We don't wrap the function itself yet
        
        return _decorator_inst
    
    def event(self, event: str | None = None):
        """Register an event.
        
        Args:
            event: The event to register to. Defaults to the name of the function."""
        
        def _decorator_inst(func: Awaitable):
            event_name = event or getattr(func, '__name__', None)
            # If the event key doesn't yet exist, create it
            if not self.events.get(event_name, False):
                self.events[event_name] = []
            self.events[event_name].append(func)
            return func # We don't wrap the function itself
        
        return _decorator_inst
    
    def start(self):
        """Start the bot and begin processing messages"""
        try:
            loop = asyncio.new_event_loop()
            loop.run_until_complete(self._start_async())    
        except Exception as e:
            print(f"Failed to start bot loop: {str(e)}")
    
    async def _start_async(self):
        """Start the bot, but in an async context."""
        try:
            print("Starting bot...")
            
            self.client_session = aiohttp.ClientSession(self.api_base_url, headers=self.headers)
            
            async with self.client_session.post(f"{self.api_path}/auth") as response:
                response_data = await response.json()
                print(f"Bot authenticated: {response_data['bot']['name']}")
        
            # Establish baseline (gets highest message ID but no messages)
            print("Establishing baseline with server...")
            
            async with self.client_session.get(f"{self.api_path}/messages", params={'cursor': 0}) as response:
                baseline_data = await response.json()
                # Save the cursor for future requests
                if baseline_data.get('cursor'):
                    self.current_cursor = baseline_data['cursor']
                    print(f"Baseline established. Starting with message ID: {self.current_cursor}")
            
            # Log command prefixes
            command_prefixes = [cmd for cmd in self.commands.keys()]
            print(f"Listening for commands: {', '.join(command_prefixes)}")
            
            await self._process_message_loop()
            
        except Exception as e:
            print(f"Failed to start bot: {str(e)}")
            traceback.print_exc()
            sys.exit(1)
    
    async def _process_message_loop(self):
        """Process messages in a loop"""
        while True:
            try:
                async with self.client_session.get(f"{self.api_path}/messages", params={'cursor': self.current_cursor}) as response:
                    data = await response.json()

                    # Update cursor
                    if data.get('cursor'):
                        self.current_cursor = data['cursor']

                    # Process each new message
                    for message in data.get('messages', []):
                        message = Message(message["id"], message["content"],
                                          message["flock_name"], message["flock_id"],
                                          message["nest_name"], message["nest_id"],
                                          datetime.strptime(message["created_at"], "%Y-%m-%dT%H:%M:%SZ"), 
                                          MessageUser(message["user"]["id"], message["user"]["username"], message["user"]["is_bot"]))
                        
                        # Run on_message_raw events
                        events = self.events.get("on_message_raw", [])
                        for event in events:
                            await event(message)
                        # Skip bot messages
                        if message.user.is_bot:
                            continue

                        print(f"Message from {message.user.username} in {message.flock_name}/{message.nest_name}: {message.content}")

                        # Check for commands
                        was_command = await self._handle_command(message)
                        # If it was a command, skip
                        if was_command:
                            continue
                        # Run on_message events
                        events = self.events.get("on_message", [])
                        for event in events:
                            await event(message)
                
                # Wait before checking again
                await asyncio.sleep(1)
                
            except Exception as e:
                print(f"Error fetching messages: {str(e)}")
                traceback.print_exc()
                await asyncio.sleep(5)  # Longer delay on error
    
    async def _handle_command(self, message: Message) -> bool:
        """
        Handle command messages
        
        Args:
            message: Message object
        """
        message_content: str = message.content.lower()
        # Check each command prefix
        for prefix, handler in self.commands.items():
            if message_content.startswith(prefix.lower()):
                print(f"Command detected: {prefix}")
                
                # Generate response using the command handler, if we don't get an error
                try:
                    response_content = await handler(message)
                except Exception as e:
                    events = self.events.get("on_command_exception", [])
                    for event in events:
                        await event(prefix, message, e)
                    return False
                    
                # Send the response
                await self.send_message(message.flock_id, message.nest_id, response_content)
                
                # Return true after first matching command
                return True
        # No commands were ran, so return false (we didn't run a command)
        return False
        
    async def send_message(self, flock_id: Union[str, int], nest_id: Union[str, int], content: str):
        """
        Send a message
        
        Args:
            flock_id: Flock ID
            nest_id: Nest ID
            content: Message content
        """
        try:
            print(f'Sending message: "{content}"')
            
            async with self.client_session.post(f"{self.api_path}/flocks/{flock_id}/nests/{nest_id}/messages", json={'content': content}) as response:
                response.raise_for_status()
                
        except Exception as e:
            print(f"Failed to send message: {str(e)}")
            
    def _handle_shutdown_stub(self, sig, frame):
        try:
            loop = asyncio.get_running_loop()
            loop.create_task(self._handle_shutdown(sig, frame))
        except RuntimeError:
            asyncio.run(self._handle_shutdown(sig, frame))

    async def _handle_shutdown(self, sig, frame):
        """Handle graceful shutdown"""
        print("Bot shutting down...")
        await self.client_session.close()
        try:
            asyncio.get_running_loop().stop()
        except RuntimeError:
            pass
        sys.exit(0)